web/pandas/pdeps/0005-no-default-index-mode.md

```diff
@@ -0,0 +1,377 @@
+
+In [13]: df.loc[df['a']>1, 'b']
+Out[13]:
+5
+6
+Name: b, dtype: int64
+
+In [14]: df.loc[df['a']>1, 'b'].index
+Out[14]: NoRowIndex(len=2)
+```
+
+Slicing by label, however, would be disallowed:
+```python
+In [15]: df.loc[0, 'b']
+---------------------------------------------------------------------------
+IndexError: Cannot use label-based indexing on NoRowIndex!
+```
+
+Note too that:
+- other uses of ``.loc``, such as boolean masks, would still be allowed (see F.A.Q);
+- ``.iloc`` and ``.iat`` would keep working as before;
+- ``.at`` would raise.
+
+### Aligning ``NoRowIndex``s
+
+To minimise surprises, the rule would be:
+
+> A ``NoRowIndex`` can only be aligned with another ``NoRowIndex`` of the same length.
+> Attempting to align it with anything else would raise.
+
+Example:
+```python
+In [1]: ser1 = pd.Series([1, 2, 3], index=NoRowIndex(3))
+
+In [2]: ser2 = pd.Series([4, 5, 6], index=NoRowIndex(3))
+
+In [3]: ser1 + ser2  # works!
+Out[3]:
+5
+7
+9
+dtype: int64
+
+In [4]: ser1 + ser2.iloc[1:]  # errors!
+---------------------------------------------------------------------------
+TypeError: Cannot join NoRowIndex of different lengths
+```
+
+### Columns cannot be NoRowIndex
+
+This proposal deals exclusively with letting users not have to think about
+row labels. There's no suggestion to remove the column labels.
+
+In particular, calling ``transpose`` on a ``NoRowIndex`` ``DataFrame``
+would error. The error would come with a helpful error message, informing
+users that they should first set an index. E.g.:
+```python
+In [4]: df = pd.DataFrame({'a': [1,  2, 3], 'b': [4, 5, 6]}, index=NoRowIndex(3))
+
+In [5]: df.transpose()
+---------------------------------------------------------------------------
+ValueError: Columns cannot be NoRowIndex.
+If you got here via `transpose` or an `axis=1` operation, then you should first set an index, e.g.: `df.pipe(lambda _df: _df.set_axis(pd.RangeIndex(len(_df))))`
+```
+
+### DataFrameFormatter and SeriesFormatter changes
+
+When printing an object with a ``NoRowIndex``, then the row labels would not be shown:
+
+```python
+In [15]: df = pd.DataFrame({'a': [1,  2, 3], 'b': [4, 5, 6]}, index=NoRowIndex(3))
+
+In [16]: df
+Out[16]:
+ a  b
+ 1  4
+ 2  5
+ 3  6
+```
+
+Of the above changes, this may be the only one that would need implementing within
+``DataFrameFormatter`` / ``SerieFormatter``, as opposed to within ``NoRowIndex``.
+
+## Usage and Impact
+
+Users would not be expected to work with the ``NoRowIndex`` class itself directly.
+Usage would probably involve a mode which would change how the ``default_index``
+function to return a ``NoRowIndex`` rather than a ``RangeIndex``.
+Then, if a user opted in to this mode with
+
+```python
+pd.set_option('mode.no_row_index', True)
+```
+
+then the following would all create a ``DataFrame`` with a ``NoRowIndex`` (as they
+all call ``default_index``):
+
+- ``df.reset_index(drop=True)``;
```

Comment:
Wouldn't `df.reset_index()` (i.e., with `drop=False`) also do the same thing?  I see no need to include `drop=True` in this example.